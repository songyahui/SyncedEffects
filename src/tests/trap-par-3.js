"use hopscript"

var hh = require("hiphop")

var prg = <hh.module S1_and_S2 S1_and_not_S2 not_S1_and_S2 not_S1_and_not_S2>
  <hh.local S1 S2>
  <hh.loop>
    <hh.trap T1>
      <hh.parallel>
	<hh.sequence>
	  <hh.emit S1/>
	  <hh.pause/>
	  <hh.exit T1/>
	</hh.sequence>
	<hh.trap T2>
	  <hh.parallel>
	    <hh.sequence>
	      <hh.emit S2/>
	      <hh.pause/>
	      <hh.exit T2/>
	    </hh.sequence>
	    <hh.sequence>
	      <hh.if S1>
		<hh.if S2>
		  <hh.emit S1_and_S2/>
		  <hh.emit S1_and_not_S2/>
		</hh.if>
		<hh.if S2>
		  <hh.emit not_S1_and_S2/>
		  <hh.emit not_S1_and_not_S2/>
		</hh.if>
	      </hh.if>
	      <hh.pause/>
	    </hh.sequence>
	  </hh.parallel>
	</hh.trap>
      </hh.parallel>
    </hh.trap>
  </hh.loop>
  </hh.local>
</hh.module>;

exports.prg = new hh.ReactiveMachine(prg, "trappar3");
