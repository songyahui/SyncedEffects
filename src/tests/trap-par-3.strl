module trappar3:
output S1_and_S2, S1_and_not_S2, not_S1_and_S2, not_S1_and_not_S2;
signal S1 in
    signal S2 in
        loop
            trap T1 in
                    [ emit S1;
                      pause;
                      exit T1 ]
                ||
                    trap T2 in
                            [ emit S2;
                              pause;
                              exit T2 ]
                        ||
                            [ present S1 then
                                  present S2 then
                                      emit S1_and_S2
                                  else
                                      emit S1_and_not_S2
                                  end present
                              else
                                  present S2 then
                                      emit not_S1_and_S2
                                  else
                                      emit not_S1_and_not_S2
                                  end present
                              end present;
                              pause ]
                    end trap
            end trap
        end loop
    end signal
end signal
end module
