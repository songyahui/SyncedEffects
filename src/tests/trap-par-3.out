trappar3> ;
--- Output: S1_and_S2
trappar3> ;
--- Output: S1_and_S2
trappar3> ;
--- Output: S1_and_S2
trappar3> ;
--- Output: S1_and_S2
